File: src/main/java/skytils/skytilsmod/utils/toasts/BlessingToast.java
package skytils.skytilsmod.utils.toasts;

import net.minecraft.client.gui.Gui;
import net.minecraft.client.renderer.GLAllocation;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.ResourceLocation;
import skytils.skytilsmod.Skytils;
import skytils.skytilsmod.utils.RenderUtil;

import java.nio.FloatBuffer;
import java.util.List;

public class BlessingToast implements IToast<BlessingToast> {
    private static final ResourceLocation TEXTURE = new ResourceLocation("skytils:gui/toast.png");
    private final FloatBuffer buffer = GLAllocation.createDirectFloatBuffer(16);
    private final long maxDrawTime;
    private final Blessing blessing;
    private final List<BlessingBuff> buffs;

    public BlessingToast(String blessing, List<BlessingBuff> buffs) {
        this.maxDrawTime = Skytils.config.toastTime;
        this.blessing = Blessing.fromName(blessing);
        this.buffs = buffs;
    }

    @Override
    public IToast.Visibility draw(GuiToast toastGui, long delta) {

        toastGui.mc.getTextureManager().bindTexture(TEXTURE);
        GlStateManager.color(1.0F, 1.0F, 1.0F);
        Gui.drawModalRectWithCustomSizedTexture(0, 0, 0, 0, 160, 32, 160, 32);

        RenderHelper.disableStandardItemLighting();

        toastGui.mc.fontRendererObj.drawStringWithShadow(this.blessing.getFormattedName(), 30, 7, 16777215);

        StringBuilder buffStats = new StringBuilder();
        for (BlessingBuff buff : this.buffs) {
            String color;
            switch (buff.symbol) {
                case "❤":
                case "❁":
                    color = EnumChatFormatting.RED.toString();
                    break;
                case "✎":
                    color = EnumChatFormatting.AQUA.toString();
                    break;
                case "❈":
                case "HP":
                    color = EnumChatFormatting.GREEN.toString();
                    break;
                case "✦":
                    color = EnumChatFormatting.WHITE.toString();
                    break;
                case "☠":
                    color = EnumChatFormatting.BLUE.toString();
                    break;
                default:
                    color = EnumChatFormatting.GRAY.toString();
            }
            buffStats.append(color).append(buff.amount).append(buff.symbol).append(" ");
        }
        GuiToast.drawSubline(toastGui, delta, 0L, this.maxDrawTime, this.buffer, buffStats.toString(), false);
        RenderHelper.enableGUIStandardItemLighting();

        RenderUtil.renderTexture(this.blessing.texture, 8, 8);

        GlStateManager.disableLighting();
        return delta >= this.maxDrawTime ? IToast.Visibility.HIDE : IToast.Visibility.SHOW;
    }

    private enum Blessing {
        LIFE("life", "skytils:toasts/blessings/life.png", "§c§lLIFE BLESSING!"),
        POWER("power", "skytils:toasts/blessings/power.png", "§5§lPOWER BLESSING!"),
        STONE("stone", "skytils:toasts/blessings/stone.png", "§a§lSTONE BLESSING!"),
        WISDOM("wisdom", "skytils:toasts/blessings/wisdom.png", "§b§lWISDOM BLESSING!"),
        TIME("time", "skytils:toasts/blessings/time.png", "§6§lTIME BLESSING!");

        private final ResourceLocation texture;
        private final String name;
        private final String formattedName;

        Blessing(String name, String location, String formattedName) {
            this.name = name;
            this.texture = new ResourceLocation(location);
            this.formattedName = formattedName;
        }

        public static Blessing fromName(String name) {
            for (Blessing type : values()) {
                if (type.name.equals(name))
                    return type;
            }
            return null;
        }

        public String getFormattedName() {
            return this.formattedName;
        }
    }

    public static class BlessingBuff {

        public String amount;
        public String symbol;

        public BlessingBuff(String amount, String symbol) {
            this.amount = amount;
            this.symbol = symbol;
        }
    }
}

File: src/main/kotlin/skytils/skytilsmod/utils/toasts/BlessingToast.kt
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2022 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package skytils.skytilsmod.utils.toasts

import net.minecraft.client.gui.Gui
import net.minecraft.client.renderer.GLAllocation
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.util.EnumChatFormatting
import net.minecraft.util.ResourceLocation
import skytils.skytilsmod.Skytils
import skytils.skytilsmod.utils.RenderUtil

class BlessingToast(blessing: String, private val buffs: List<BlessingBuff>) : IToast<BlessingToast> {
    private val buffer = GLAllocation.createDirectFloatBuffer(16)
    private val maxDrawTime: Long = Skytils.config.toastTime.toLong()
    private val blessing: Blessing? = Blessing.fromName(blessing)
    override fun draw(toastGui: GuiToast, delta: Long): IToast.Visibility {
        toastGui.mc.textureManager.bindTexture(TEXTURE)
        GlStateManager.color(1.0f, 1.0f, 1.0f)
        Gui.drawModalRectWithCustomSizedTexture(0, 0, 0f, 0f, 160, 32, 160f, 32f)
        RenderHelper.disableStandardItemLighting()
        toastGui.mc.fontRendererObj.drawStringWithShadow(blessing!!.formattedName, 30f, 7f, 16777215)
        val buffStats = StringBuilder()
        for (buff in buffs) {
            val color: String = when (buff.symbol) {
                "❤", "❁" -> EnumChatFormatting.RED.toString()
                "✎" -> EnumChatFormatting.AQUA.toString()
                "❈", "HP" -> EnumChatFormatting.GREEN.toString()
                "✦" -> EnumChatFormatting.WHITE.toString()
                "☠" -> EnumChatFormatting.BLUE.toString()
                else -> EnumChatFormatting.GRAY.toString()
            }
            buffStats.append(color).append(buff.amount).append(buff.symbol).append(" ")
        }
        GuiToast.drawSubline(toastGui, delta, 0L, maxDrawTime, buffer, buffStats.toString(), false)
        RenderHelper.enableGUIStandardItemLighting()
        RenderUtil.renderTexture(blessing.texture, 8, 8)
        GlStateManager.disableLighting()
        return if (delta >= maxDrawTime) IToast.Visibility.HIDE else IToast.Visibility.SHOW
    }

    private enum class Blessing(val blessingName: String, location: String, val formattedName: String) {
        LIFE("life", "skytils:toasts/blessings/life.png", "§c§lLIFE BLESSING!"), POWER(
            "power",
            "skytils:toasts/blessings/power.png",
            "§5§lPOWER BLESSING!"
        ),
        STONE("stone", "skytils:toasts/blessings/stone.png", "§a§lSTONE BLESSING!"), WISDOM(
            "wisdom",
            "skytils:toasts/blessings/wisdom.png",
            "§b§lWISDOM BLESSING!"
        ),
        TIME("time", "skytils:toasts/blessings/time.png", "§6§lTIME BLESSING!");

        val texture: ResourceLocation = ResourceLocation(location)

        companion object {
            fun fromName(name: String): Blessing? {
                for (type in values()) {
                    if (type.blessingName == name) return type
                }
                return null
            }
        }

    }

    class BlessingBuff(var amount: String, var symbol: String)
    companion object {
        private val TEXTURE = ResourceLocation("skytils:gui/toast.png")
    }

}